### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by MohameDawood**
we used the minimax algorithm with alpha-beta enhancement

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    }
+  end
+
+  attr_accessor :board
+
+  attr_reader :player, :converter
+
+  def display
+    puts " #{board[0][0]} | #{board[0][1]} | #{board[0][2]}"
```
Comment (arslanbisharat) [inline lib/board.rb:51]: - [ ] Please remove puts. 
only_main_print

File: lib/board.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    puts '-----------'
```
Comment (arslanbisharat) [inline lib/board.rb:53]: - [ ] Please remove puts. 
only_main_print

File: lib/board.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+    puts " #{board[1][0]} | #{board[1][1]} | #{board[1][2]}"
+
+    puts '-----------'
+
```
Comment (arslanbisharat) [inline lib/board.rb:58]: - [ ] Please remove puts. 
only_main_print

File: lib/board.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  attr_accessor :board
+
+  attr_reader :player, :converter
+
+  def display
+    puts " #{board[0][0]} | #{board[0][1]} | #{board[0][2]}"
+
+    puts '-----------'
+
+    puts " #{board[1][0]} | #{board[1][1]} | #{board[1][2]}"
```
Comment (arslanbisharat) [inline lib/board.rb:55]: - [ ] Please remove puts. 
only_main_print

File: lib/board.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    puts '-----------'
+
+    puts " #{board[2][0]} | #{board[2][1]} | #{board[2][2]}"
```
Comment (arslanbisharat) [inline lib/board.rb:59]: - [ ] Please remove puts. 
only_main_print

File: lib/board.rb [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  end
+
+  def play
+    result = ''
+
+    puts "\n\nTIC TAC TIE"
```
Comment (arslanbisharat) [inline lib/board.rb:65]: - [ ] Please remove puts. 
only_main_print

File: lib/board.rb [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+      puts "\nInsert Position"
+
+      number = gets.chomp.to_i
+
+      if board[converter[number][0]][converter[number][1]] == 'o' ||
+         board[converter[number][0]][converter[number][1]] == 'x'
+
+        puts 'This Position is Occupied Choice another one'
+        number = gets.chomp.to_i
```
Comment (arslanbisharat) [inline lib/board.rb:81]: - [ ] Please remove puts and gets.chomp. 
only_main_print

**Review by arslanbisharat — CHANGES_REQUESTED** [review]
### Changes requested 😸

**Your project is almost done please work on these changes.**

- [x] No `puts/print `and` get.chomp` outside main.rb file.
only_main_print

- [x] When the entry is invalid ( not in the range of 1-9) it should say something like "invalid entry please try between 1-9". But in your app, it gives an undefined error. Please work on it. 


- [x]  From your board, position 1 is not the first cell on the board, it's somewhere in the end. Fix that.

a

- [x]  The application should not allow a player to move into an occupied position. Your game only gives the warning but it still allows the user to occupy the position.

b

- [x] The application does not exist/stop due to invalid input - it informs the player about it and asks for their move again

c

 - [x] The project uses at least two classes (Usually it uses three: Game, Board, and Player). The code of the project should be logically divided between classes.

  - [x] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). `erere.rb` and `board.rb` contains similar codes.

  - [x] DRY rule should be respected. E.g. students tend to check winning position for X and O separately instead of creating a method and reuse it

- [x] The application should tell if there was a draw after players move. Your application is flawed, both moves are winning moves, fix this.

d

In case of any question please reach out to slack: arslanbisharat

**Best,
Arslan Bisharat**

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
## Changes required #
**Hi @MohameDawood & @daviidy**  
please make sure to do the changes required by the previous tse. if you did so, please make sure to push the code before submitting another request.

**Comment by mimipeshy** [comment]
## STATUS: CHANGES REQUIRED :red_circle: 

Hi team, you are doing a great job so far, your project looks promising but we have to change a few things to make it work perfectly. All required changes are **MANDATORY**

- [ ] As mentioned by the two TSE's, No puts/print and get.chomp outside main.rb file.([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)). 

- [ ] Please give users an option to input their names e.g player 1 and player 2 and assign a symbol, your game will look more user-friendly

**Behavior**

- [ ] From your board, position 1 is not the first cell on the board, it's somewhere in the middle. Fix that.
Screenshot from 2020-10-30 11-11-42


- [ ] The application does not allow a player to move into an occupied position. Your game only gives the warning but it still allows the user to occupy the position.

- When the user chooses a position, both symbols appear on the board. Fix that. Only one user is allowed to play at a time.

Screenshot from 2020-10-30 10-58-48

- [ ] The application does not exit/stop due to invalid input - it informs the player about it and asks for their move again
Screenshot from 2020-10-30 11-03-45

- [ ]  The application can tell if there was a draw after players move. Your application is flawed, both moves are winning moves, fix this.
Screenshot from 2020-10-30 11-09-33

- [ ] The application can change active player. As a user, I am playing blindly, I don't know whose turn it is. This can be implemented by applying user names as I stated in the second point.

**Implementation details**

- [ ] The project uses at least two classes (Usually it uses three: Game, Board, and Player). The code of the project should be logically divided between classes.

- [ ] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). erere.rb and board.rb contains similar codes.

- [ ] DRY rule should be respected. E.g. students tend to check winning position for X and O separately instead of creating a method and reuse it. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/dry.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

```ruby
class Game
  board = ['O', 'X', 'O', 'X', 'O', 'X', 'O', 'X', 'X']

  # bad
  def check_win
    # for rows
    return 'X' if board[0] == board[1] == board[2] == 'X'
    return 'O' if board[0] == board[1] == board[2] == 'O'
    return 'X' if board[3] == board[4] == board[5] == 'X'
    return 'O' if board[3] == board[4] == board[5] == 'O'
    return 'X' if board[6] == board[7] == board[8] == 'X'
    return 'O' if board[6] == board[7] == board[8] == 'O'

    # and analogical as above for columns and diagonals
    # which makes it really long...

  end

  # better
  def check_win(player_symbol)
    # for rows
    all_equal(board[0..2], player_symbol) ||
    all_equal(board[3..5], player_symbol) ||
    all_equal(board[6..8], player_symbol)

    # and analogical as above for columns and diagonals
  end

  def all_equal(subarray, player_symbol)
    return true if subarray.uniq.size == 1 && subarray.uniq == player_symbol
  end
end
```
 
That's it :relieved: , please work on ALL changes and submit your work for review. :smile: 
Happy Coding :computer: 

Peris Ndanu
Reach out to me on slack in case of any questions (peris ndanu)

**Pull request merged**